src/ripple/app/ledger/impl/TransactionMaster.cpp

```diff
@@ -44,22 +44,24 @@ bool TransactionMaster::inLedger (uint256 const& hash, std::uint32_t ledger)
     return true;
 }
 
-std::shared_ptr<Transaction>
-TransactionMaster::fetch (uint256 const& txnID, bool checkDisk)
+auto
+TransactionMaster::fetch (uint256 const& txnID, bool checkDisk) -> TransactionEnvelope
 {
-    auto txn = mCache.fetch (txnID);
+    TransactionEnvelope ret;
 
-    if (!checkDisk || txn)
-        return txn;
+    ret.txn = mCache.fetch (txnID);
 
-    txn = Transaction::load (txnID, mApp);
+    if (!checkDisk || ret.txn)
+        return ret;
```

Comment:
If the transaction is in the cache we don't return the range but if it's not we do? That seems strange.

Reply:
Yeah, it seems strange initially, but the reason for this is because we don't need to report the range if the transaction is found. The client only needs the range when the transaction is not found in order to determine that the returned range of ledgers do not contain the requested transaction.

Reply:
If we only return the searched range when it's not found, what do you think about using a `std::variant` as the return type? I like that this makes the API clear: we either return the transaction or the searched range. This also solves renaming the `TransactionEnvelope` struct.

I'll admit I'm not a huge fan of `std::variant`. It's cumbersome to use. However in this case it _might_ work out OK.

Reply:
Reverted this file.